File: as04.py
# as04.py
# Evaluates the truth of a given propositional statement in postfix notation. 

from itertools import product

def findProps(tokens):
    props = set()
    for token in tokens:
        if (token is None):
            print("Empty string detected.")
        elif (token == "="):
            pass
        elif (token == "->"):
            pass
        elif (token == "<="):
            pass
        elif (token == "+"):
            pass
        elif (token == "*"):
            pass
        elif (token == "!"):
            pass
        elif (token == "~"):
            pass
        elif (token == "True"):
            pass
        elif (token == "False"):
            pass
        else:
            props.add(token)
    return props

def calculateTruth(propMap, premise):
    operands = list()
    numOfPushes = 0
    st = list()

    for token in premise:
        if (token is None):
            print("Empty string detected.")
        elif (token == "="):
            operands.clear();
            operands.append(st.pop())
            operands.append(st.pop())
            st.append(operands[0] is operands[1])
        elif (token == "->"):
            operands.clear();
            operands.append(st.pop())
            operands.append(st.pop())
            st.append(operands[0] or not operands[1])
        elif (token == "<="):
            operands.clear();
            operands.append(st.pop())
            operands.append(st.pop())
            st.append(operands[0] or not operands[1])
        elif (token == "+"):
            operands.clear();
            operands.append(st.pop())
            operands.append(st.pop())
            st.append(operands[0] or operands[1])
        elif (token == "*"):
            operands.clear();
            operands.append(st.pop())
            operands.append(st.pop())
            st.append(operands[0] and operands[1])
        elif (token == "!"):
            operands.clear();
            operands.append(st.pop())
            st.append(not operands[0])
        elif (token == "~"):
            operands.clear();
            operands.append(st.pop())
            st.append(not operands[0])
        elif (token == "True"):
            operands.clear();
            operands.append(True)
            st.append(True)
        elif (token == "False"):
            operands.clear();
            operands.append(True)
            st.append(True)
        else:
            st.append(propMap.get(token))
            numOfPushes += 1
            pass

    truthValue = st.pop()
    return truthValue;

def generateTable(sortedPropList, premises, conclusion):
    propMap = dict()

    for row in product((False, True), repeat=len(sortedPropList)):
        premiseTruthList = list()
        truth = list()
        for item in row:
            truth.append(item)
        for proposition, i in zip(sortedPropList, range(len(truth))):
            propMap[proposition] = truth[i]
        for premise in premises:
            premiseTruthList.append(calculateTruth(propMap, premise.strip().split(" ")))
            conclusionTruth = calculateTruth(propMap, conclusion.strip().split(" "))
        if (all(item == True for item in premiseTruthList) and conclusionTruth == False):
            return False
    return True

inputList = list()
propSet = set()
while True:
    try:
        line = input()
    except EOFError:
        break
    inputList.append(line)
if(len(inputList) > 0):
    conclusion = inputList[-1:][0]
    premises = (inputList[:-1])
    for item in premises:
        propSet.update(findProps(item.strip().split(" ")))
    sortedPropList = list(sorted(propSet))
    print(generateTable(sortedPropList, premises, conclusion))

File: as04.java
import java.util.*;

/**
 *
 * @author tanner yost
 */
public class TruthTables {
    /**
     * method getLength returns the number of propositions in the statement
     * @param tokens
     * @return
     */
    private static int GetLength(Set<String> tokens) {
        int length = 0;
        for (String token : tokens) {
            if (null == token) {
                System.out.println("Empty string detected.");
            }
            else switch (token) {
                case "=":
                    break;
                case "->":
                    break;
                case "<=":
                    break;
                case "+":
                    break;
                case "*":
                    break;
                case "!":
                    break;
                case "~":
                    break;
                case "True":
                    break;
                case "False":
                    break;
                default:
                    length++;
                    break;
            }
        }

        return length;
    }

    /**
     * method getInput returns the provided input for later use
     * @return
     */
    private static String[] GetInput() {

        char[] tokens;
        Scanner sc = new Scanner(System.in);
        Set<String> inputSet = new HashSet<>();

        String input = sc.nextLine();
        //String input = "p p * r *";
        String expression[] = input.split(" ");
        String[] trimmedArray = new String[expression.length];
        for (int i = 0; i < expression.length; i++) {
            trimmedArray[i] = expression[i].trim();
        }

        return trimmedArray;

    }
    private static Set createSet(String[] a) {
        Set<String> inputSet = new HashSet<>();

        for(String token : a){
            inputSet.add(token);
        }
        return inputSet;
    }

    /**
     * method print truth table will provide a truth table, based on the number of propositions.
     * It will then populate the table, for each propositions it will provide a true/false value.
     * Then it will send these values into calculateTruth which will print out on the last column the truth value of the entire statement
     * @param n, length
     * @param st
     * @param propositions
     */
    private static void PrintTruthTable(int n, Set<String> input, String[] inputString) {
        int rows = (int) Math.pow(2,n);
        HashMap propositions;
        for (int i=0; i<rows; i++) { // for each row
            boolean[] truth = new boolean[n];
            for (int j=n-1; j>=0; j--) {
                int value = ((i/(int) Math.pow(2, j))%2);
                if(value == 1){
                    truth[j] = true;
                    System.out.print(truth[j]);
                    System.out.print("\t");
                }
                else if (value == 0){
                    truth[j] = false;
                    System.out.print(truth[j]);
                    System.out.print("\t");
                }
            }

            propositions = LinkMap(truth, n, input);
            System.out.println(CalculateTruth(n, propositions, inputString));
            //System.out.println(propositions);
        }
    }
    /**
     * method calculateTruth returns the truth value
     * @param st
     * @param propositions
     */
    private static Boolean CalculateTruth(int length, HashMap propositions, String[] input){
        ArrayList<Boolean> operands = new ArrayList<>();
        int numberOfPushes = 0;
        Stack<Boolean> st = new Stack<>();

        for (String token : input) {
            if (null == token) {
                System.out.println("Empty string detected.");
            }
            else switch (token) {
                case "=":
                    operands.clear();
                    operands.add(st.pop());
                    operands.add(st.pop());
                    st.push(operands.get(0) == operands.get(1));
                    break;
                case "->":
                    operands.clear();
                    operands.add(st.pop());
                    operands.add(st.pop());
                    st.push(operands.get(0) || !operands.get(1));
                    break;
                case "<=":
                    operands.clear();
                    operands.add(st.pop());
                    operands.add(st.pop());
                    st.push(operands.get(0) || !operands.get(1));
                    break;
                case "+":
                    operands.clear();
                    operands.add(st.pop());
                    operands.add(st.pop());
                    st.push(operands.get(0) || operands.get(1));
                    break;
                case "*":
                    operands.clear();
                    operands.add(st.pop());
                    operands.add(st.pop());
                    st.push(operands.get(0) && operands.get(1));
                    break;
                case "!":
                    operands.clear();
                    operands.add(st.pop());
                    st.push(!operands.get(0));
                    break;
                case "~":
                    operands.clear();
                    operands.add(st.pop());
                    st.push(!operands.get(0));
                    break;
                case "True":
                    operands.clear();
                    operands.add(true);
                    st.push(true);
                    break;
                case "False":
                    operands.clear();
                    operands.add(false);
                    st.push(false);
                    break;
                default:
                    st.push((Boolean) propositions.get(token));
                    numberOfPushes++;
                    break;
            }
        }
        Boolean truthValue = st.pop();
        return truthValue;
    }
    /**
     * method linkMap will create a hashmap for the truth values of the propositions
     */
    private static HashMap LinkMap(boolean[] truth, int n, Set<String> input) {
        HashMap<String, Boolean> propositions = new HashMap<>();

        int j = n-1;
        System.out.println(j);
        for (String token : input) {
                if (null == token) {
                    System.out.println("Empty string detected.");
                }
                else switch (token) {
                    case "=":
                        break;
                    case "->":
                        break;
                    case "<=":
                        break;
                    case "+":
                        break;
                    case "*":
                        break;
                    case "!":
                        break;
                    case "~":
                        break;
                    case "True":
                        break;
                    case "False":
                        break;
                    default:
                        propositions.put(token, truth[j]);
                        j--;
                        break;
                }
        }

        return propositions;
    }
    /**
     * Prints the first row of the truth table, the header.
     * Contains all of the propositions to be examined as well as a column for the whole statement.
     * @param length, integer representing how many propositions there are
     * @param input, the string representing the entire input. Separated by whitespace and then printed.
     */
    private static void PrintHeader(int length, Set<String> input, String[] inputString) {
        List<String> sortedList = new ArrayList<>();


        if(length>0) {
            for (String token : input) {
                if (null == token) {
                    System.out.println("Empty string detected.");
                }
                else switch (token) {
                    case "=":
                        break;
                    case "->":
                        break;
                    case "<=":
                        break;
                    case "+":
                        break;
                    case "*":
                        break;
                    case "!":
                        break;
                    case "~":
                        break;
                    case "True":
                        break;
                    case "False":
                        break;
                    default:
                        sortedList.add(token);
                        break;
                }
            }

            Collections.sort(sortedList);
            for(String a : sortedList){
                System.out.print(a);
                System.out.print("\t");
            }

            for(String element : inputString){
                System.out.print(element);
                System.out.print(" ");
            }
            System.out.println("");
        }

    }
    public static void main(String[] args) {
        Set<String> inputSet = new HashSet<>();
        String[] input = GetInput();
        inputSet = createSet(input);
        int length = GetLength(inputSet);
        System.out.println(length);
        PrintHeader(length, inputSet, input);
        PrintTruthTable(length, inputSet, input);
    }
}
